src/common/core/librarycore.js

```diff
@@ -0,0 +1,785 @@
+                    libraryRoot = getLibraryRoot(node);
+                }
+                ASSERT(self.isValidNode(libraryRoot) &&
+                    getLibraryName(node).indexOf(getLibraryName(libraryRoot) === 0));
+
+                if (self.isLibraryRoot(node) && self.getPath(node) === self.getPath(libraryRoot)) {
+                    return innerCore.getDataGuid(node);
+                }
+
+                if (!self.isLibraryElement(node)) {
+                    return null; //normal nodes do not have library GUID
+                }
+
+                return innerCore.getDeductedGuid(node, self.getLibraryGuid(self.getParent(node), libraryRoot));
+            };
+
+            this.addLibrary = function (node, name, libraryRootHash, libraryInfo) {
+                var root = self.getRoot(node),
+                    libraryRelid = RANDOM.generateRelid(root.data);
+
+                innerCore.setProperty(root, libraryRelid, libraryRootHash);
+                return TASYNC.call(function (newLibraryRoot) {
+                    return TASYNC.call(function (libraryNodes) {
+                        var inMeta = self.getMemberPaths(newLibraryRoot, CONSTANTS.META_SET_NAME),
+                            libraryInfoAttribute = libraryInfo;
+                        //remove the libraryRoot from the libraryNodes
+                        libraryNodes.shift();
+
+                        //set the name of the library root
+                        innerCore.setAttribute(newLibraryRoot, 'name', name);
+
+                        //add library_info
+                        libraryInfoAttribute.hash = libraryRootHash;
+                        innerCore.setAttribute(newLibraryRoot, '_libraryInfo', libraryInfoAttribute);
+
+                        if (libraryNodes.length > 0) {
+                            //connect the FCO as base of libraryFCO
+                            innerCore.setBase(self.getBaseRoot(libraryNodes[0]), self.getFCO(root));
+
+                            //adding nodes to the global META
+                            var i;
+                            for (i = 0; i < libraryNodes.length; i += 1) {
+                                if (inMeta.indexOf(self.getPath(libraryNodes[i])) !== -1) {
+                                    innerCore.addMember(root, CONSTANTS.META_SET_NAME, libraryNodes[i]);
+                                }
+                            }
+                        }
+                        //refreshing libraryInfo
+                        root.libraryRoots[name] = newLibraryRoot;
+                    }, self.loadSubTree(newLibraryRoot));
+                }, self.loadChild(root, libraryRelid));
+            };
+
+            this.updateLibrary = function (node, name, updatedLibraryRootHash, libraryInfo, updateInstructions) {
+                var logs = {added: {}, updated: {}, moved: {}, removed: {}},
+                    root = self.getRoot(node),
+                    libraryRoot = getRootOfLibrary(root, name),
```

Comment:
getLibraryRootByName could be confused with getLibraryRoot, because both function would need a node to get access to the containment hierarchy. I know it is not ideal, but I hope that at least the public API function names are straigthforward